File: DL-TP/DL-TP.cs
using System;
using System.IO;
using System.Collections.Generic;

public class DecentlabDecoder
{
  private delegate double conversion(double[] x);
  public const int PROTOCOL_VERSION = 2;

  private class Sensor
  {
    internal int length { get; set; }
    internal List<SensorValue> values { get; set; }
    internal Sensor(int length, List<SensorValue> values)
    {
      this.length = length;
      this.values = values;
    }
  }

  private class SensorValue
  {
    internal string name { get; set; }
    internal string unit { get; set; }
    internal conversion convert;
    internal SensorValue(string name, string unit, conversion convert)
    {
      this.name = name;
      this.unit = unit;
      this.convert = convert;
    }
  }

  private static readonly List<Sensor> SENSORS = new List<Sensor>() {
    new Sensor(16, new List<SensorValue>() {
      new SensorValue("CH0: Temperature", "°C", x => (x[0] - 32768) / 100),
      new SensorValue("CH1: Temperature", "°C", x => (x[1] - 32768) / 100),
      new SensorValue("CH2: Temperature", "°C", x => (x[2] - 32768) / 100),
      new SensorValue("CH3: Temperature", "°C", x => (x[3] - 32768) / 100),
      new SensorValue("CH4: Temperature", "°C", x => (x[4] - 32768) / 100),
      new SensorValue("CH5: Temperature", "°C", x => (x[5] - 32768) / 100),
      new SensorValue("CH6: Temperature", "°C", x => (x[6] - 32768) / 100),
      new SensorValue("CH7: Temperature", "°C", x => (x[7] - 32768) / 100),
      new SensorValue("CH8: Temperature", "°C", x => (x[8] - 32768) / 100),
      new SensorValue("CH9: Temperature", "°C", x => (x[9] - 32768) / 100),
      new SensorValue("CH10: Temperature", "°C", x => (x[10] - 32768) / 100),
      new SensorValue("CH11: Temperature", "°C", x => (x[11] - 32768) / 100),
      new SensorValue("CH12: Temperature", "°C", x => (x[12] - 32768) / 100),
      new SensorValue("CH13: Temperature", "°C", x => (x[13] - 32768) / 100),
      new SensorValue("CH14: Temperature", "°C", x => (x[14] - 32768) / 100),
      new SensorValue("CH15: Temperature", "°C", x => (x[15] - 32768) / 100)
    }),
    new Sensor(1, new List<SensorValue>() {
      new SensorValue("Battery voltage", "V", x => x[0] / 1000)
    })
  };

  private static int ReadInt(Stream stream)
  {
    return (stream.ReadByte() << 8) + stream.ReadByte();
  }


  public static Dictionary<string, Tuple<double, string>> Decode(byte[] msg)
  {
    return Decode(new MemoryStream(msg));
  }

  public static Dictionary<string, Tuple<double, string>> Decode(String msg)
  {
    byte[] output = new byte[msg.Length / 2];
    for (int i = 0, j = 0; i < msg.Length; i += 2, j++)
    {
      output[j] = (byte)int.Parse(msg.Substring(i, 2), System.Globalization.NumberStyles.HexNumber);
    }
    return Decode(output);
  }

  public static Dictionary<string, Tuple<double, string>> Decode(Stream msg)
  {
    var version = msg.ReadByte();
    if (version != PROTOCOL_VERSION)
    {
      throw new InvalidDataException("protocol version " + version + " doesn't match v2");
    }

    var result = new Dictionary<string, Tuple<double, string>>();
    result["Protocol version"] = new Tuple<double, string>(version, null);

    var deviceId = ReadInt(msg);
    result["Device ID"] = new Tuple<double, string>(deviceId, null);

    var flags = ReadInt(msg);
    foreach (var sensor in SENSORS)
    {
      if ((flags & 1) == 1)
      {
        double[] x = new double[sensor.length];
        for (int i = 0; i < sensor.length; i++)
        {
          x[i] = ReadInt(msg);
        }
        foreach (var val in sensor.values)
        {
          if (val.convert != null)
          {
            result[val.name] = new Tuple<double, string>(val.convert(x), val.unit);
          }
        }
      }
      flags >>= 1;
    }

    return result;
  }
}
public class Program
{
  public static void Main()
  {
    var payloads = new string[] {
      "023e3e00038abc8a928aa08a848ab38a898ac38aad8ab78a928aa1000000000000000000000afc",
      "023e3e00020afc"
    };

    foreach (var pl in payloads)
    {
      var decoded = DecentlabDecoder.Decode(pl);
      foreach (var k in decoded.Keys)
      {
        Console.WriteLine(k + ": " + decoded[k]);
      }
      Console.WriteLine();
    }
  }
}

File: DL-TP/DL-TP.js
var decentlab_decoder = {
  PROTOCOL_VERSION: 2,
  SENSORS: [
    {length: 16,
     values: [{name: 'ch0_temperature',
               displayName: 'CH0: Temperature',
               convert: function (x) { return (x[0] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch1_temperature',
               displayName: 'CH1: Temperature',
               convert: function (x) { return (x[1] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch2_temperature',
               displayName: 'CH2: Temperature',
               convert: function (x) { return (x[2] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch3_temperature',
               displayName: 'CH3: Temperature',
               convert: function (x) { return (x[3] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch4_temperature',
               displayName: 'CH4: Temperature',
               convert: function (x) { return (x[4] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch5_temperature',
               displayName: 'CH5: Temperature',
               convert: function (x) { return (x[5] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch6_temperature',
               displayName: 'CH6: Temperature',
               convert: function (x) { return (x[6] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch7_temperature',
               displayName: 'CH7: Temperature',
               convert: function (x) { return (x[7] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch8_temperature',
               displayName: 'CH8: Temperature',
               convert: function (x) { return (x[8] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch9_temperature',
               displayName: 'CH9: Temperature',
               convert: function (x) { return (x[9] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch10_temperature',
               displayName: 'CH10: Temperature',
               convert: function (x) { return (x[10] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch11_temperature',
               displayName: 'CH11: Temperature',
               convert: function (x) { return (x[11] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch12_temperature',
               displayName: 'CH12: Temperature',
               convert: function (x) { return (x[12] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch13_temperature',
               displayName: 'CH13: Temperature',
               convert: function (x) { return (x[13] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch14_temperature',
               displayName: 'CH14: Temperature',
               convert: function (x) { return (x[14] - 32768) / 100; },
               unit: '°C'},
              {name: 'ch15_temperature',
               displayName: 'CH15: Temperature',
               convert: function (x) { return (x[15] - 32768) / 100; },
               unit: '°C'}]},
    {length: 1,
     values: [{name: 'battery_voltage',
               displayName: 'Battery voltage',
               convert: function (x) { return x[0] / 1000; },
               unit: 'V'}]}
  ],

  read_int: function (bytes, pos) {
    return (bytes[pos] << 8) + bytes[pos + 1];
  },

  decode: function (msg) {
    var bytes = msg;
    var i, j;
    if (typeof msg === 'string') {
      bytes = [];
      for (i = 0; i < msg.length; i += 2) {
        bytes.push(parseInt(msg.substring(i, i + 2), 16));
      }
    }

    var version = bytes[0];
    if (version != this.PROTOCOL_VERSION) {
      return {error: "protocol version " + version + " doesn't match v2"};
    }

    var deviceId = this.read_int(bytes, 1);
    var flags = this.read_int(bytes, 3);
    var result = {'protocol_version': version, 'device_id': deviceId};
    // decode payload
    var pos = 5;
    for (i = 0; i < this.SENSORS.length; i++, flags >>= 1) {
      if ((flags & 1) !== 1)
        continue;

      var sensor = this.SENSORS[i];
      var x = [];
      // convert data to 16-bit integer array
      for (j = 0; j < sensor.length; j++) {
        x.push(this.read_int(bytes, pos));
        pos += 2;
      }

      // decode sensor values
      for (j = 0; j < sensor.values.length; j++) {
        var value = sensor.values[j];
        if ('convert' in value) {
          result[value.name] = {displayName: value.displayName,
                                value: value.convert.bind(this)(x)};
          if ('unit' in value)
            result[value.name]['unit'] = value.unit;
        }
      }
    }
    return result;
  }
};

function main() {
  console.log(decentlab_decoder.decode("023e3e00038abc8a928aa08a848ab38a898ac38aad8ab78a928aa1000000000000000000000afc"));
  console.log(decentlab_decoder.decode("023e3e00020afc"));
}

main();
